vnext/Mso/src/dispatchQueue/threadPoolScheduler_win.cpp

```diff
@@ -42,12 +50,29 @@ struct ThreadPoolSchedulerWin : Mso::UnknownObject<IDispatchQueueScheduler> {
   };
 
  private:
-  std::unique_ptr<TP_WORK, ThreadPoolWorkDeleter> m_threadPoolWork;
+  std::shared_ptr<TP_WORK> m_threadPoolWork;
   Mso::WeakPtr<IDispatchQueueService> m_queue;
   const uint32_t m_maxThreads{1};
   std::atomic<uint32_t> m_usedThreads{0};
 
   constexpr static uint32_t MaxConcurrentThreads{64};
+
+ private:
+  static std::mutex s_threadPoolWorkMutex;
+  static bool s_enableThreadPoolWorkTracking;
+  static std::vector<std::shared_ptr<TP_WORK>> s_trackedThreadPoolWork;
+};
+
+// Track the ThreadPoolSchedulerWin instance used by current thread.
+// We use it to avoid a deadlock on queue shutdown.
+struct ThreadPoolSchedulerWinContext {
+  ThreadPoolSchedulerWinContext(ThreadPoolSchedulerWin *scheduler) noexcept;
+  ~ThreadPoolSchedulerWinContext() noexcept;
+  static ThreadPoolSchedulerWin *CurrentScheduler() noexcept;
```

Comment:
is it ok to use the same threadpool context for different experiences (potentially running in separate instances) in the same process?

Reply:
> is it ok to use the same threadpool context for different experiences (potentially running in separate instances) in the same process?

Yes. This thread pool context is used only inside of the thread pool work invocation. Other queues cannot do work on the same thread at that time. The system may have multiple instances or dispatch queues. They all use different threads at any point of time.
This context is only to determine that if a specific thread at this point runs tasks from the specified scheduler. It helps preventing the deadlock because in such case the scheduler is not allowed to wait for the running work to finish.